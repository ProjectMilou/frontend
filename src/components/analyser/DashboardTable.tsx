import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  lighten,
  ListItemText,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Theme,
  Typography,
} from '@material-ui/core';
import classNames from 'classnames';
import * as API from '../../analyser/APIClient';
import EuroCurrency from './EuroCurrency';
import Performance from './Performance';

const useStyles = makeStyles((theme: Theme) => ({
  action: { display: 'inline-block' },
  row: {
    cursor: 'pointer',
  },
  rowHover: {
    backgroundColor: lighten(theme.palette.primary.light, 0.85),
  },
  positionCount: {
    fontSize: '24px',
  },
  disabled: {
    cursor: 'not-allowed',
  },
}));

export type DashboardTableRowProps = {
  stock: API.Stock;
};

export const DashboardTableRow: React.FC<DashboardTableRowProps> = ({
  stock
}) => {
  const [hover, setHover] = React.useState<boolean>(false);

  const { t } = useTranslation();
  const classes = useStyles();

  const selectStock = (symbol: API.Stock['symbol']) => {
    console.log(symbol)
  }

  return (
    <TableRow
      onClick={() => {
        selectStock(stock.symbol);
      }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      className={classNames(classes.row, hover && classes.rowHover)}
    >
      <TableCell align="center">{stock.name}</TableCell>
      <TableCell align="center" component="th" scope="row">
        <ListItemText
          primary={<Typography color="primary">{stock.symbol}</Typography>}
          secondary={
           stock['1d'] > 0 ? t('portfolio.virtual') : t('portfolio.real')
          }
          secondaryTypographyProps={{
            color: stock['1d'] > 0 ? 'textSecondary' : 'secondary',
          }}
        />
      </TableCell>
      <TableCell align="center" className={classes.positionCount}>
        {stock.analystTargetPrice}
      </TableCell>
      <TableCell align="center">
        <EuroCurrency value={stock.valuation} />
      </TableCell>
      <TableCell align="center">
        <Performance value={stock['7d']} />
      </TableCell>
      <TableCell align="center">
        <Performance value={stock['30d']} />
      </TableCell>
    </TableRow>
  );
};

export type DashboardTableProps = {
  stocks: API.Stock[];
};

const DashboardTable: React.FC<DashboardTableProps> = ({
  stocks
}) => {
  const { t } = useTranslation();

  // TODO: Improve portfolio score visualization
  return (
    <TableContainer component={Paper}>
      <Table aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell align="center">{t('portfolio.score')}</TableCell>
            <TableCell align="center">{t('portfolio.name')}</TableCell>
            <TableCell align="center">
              {t('portfolio.positionsCount')}
            </TableCell>
            <TableCell align="center">{t('portfolio.value')}</TableCell>
            <TableCell align="center">{t('portfolio.7d')}</TableCell>
            <TableCell align="center">{t('portfolio.1y')}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {stocks.map((s) => (
            <DashboardTableRow
              stock={s}
              key={s.symbol}
            />
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default DashboardTable;
