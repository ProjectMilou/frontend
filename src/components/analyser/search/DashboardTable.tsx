// Based on Portfolio's DashboardTable.tsx Will be later either replaced by Material-UI list or refactored

import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  lighten,
  ListItemText,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Theme,
  Typography,
  createStyles,
} from '@material-ui/core';
import classNames from 'classnames';
import * as API from '../../../analyser/APIClient';
import EuroCurrency from '../../shared/EuroCurrency';
import Performance from '../../shared/Performance';
import Valuation from '../../shared/Valuation';

const useStyles = makeStyles(({ palette }: Theme) =>
  createStyles({
    action: { display: 'inline-block' },
    row: {
      cursor: 'pointer',
    },
    rowHover: {
      backgroundColor: lighten(palette.primary.light, 0.85),
    },
    defaultText: {
      fontSize: '24px',
      color: palette.primary.main,
    },
    disabled: {
      cursor: 'not-allowed',
    },
  })
);

export type DashboardTableRowProps = {
  stock: API.Stock;
  selectStock: (id: string) => void;
};

// Rounds and adds M=Million, B=Billion and K=Thousand --> American System!!!
function moneyFormat(val: number): string {
  let round = '';
  if (Math.abs(val) >= 1.0e9) {
    round = `€${Math.round(Math.abs(val) / 1.0e9)}B`;
  } else if (Math.abs(val) >= 1.0e6) {
    round = `€${Math.round(Math.abs(val) / 1.0e6)}M`;
  } else if (Math.abs(val) >= 1.0e3) {
    round = `€${Math.round(Math.abs(val) / 1.0e3)}K`;
  } else {
    round = `€${Math.abs(val)}`;
  }
  return round;
}

// Numbers are output in different colors need clear approach!
export const DashboardTableRow: React.FC<DashboardTableRowProps> = ({
  stock,
  selectStock,
}) => {
  const [hover, setHover] = React.useState<boolean>(false);

  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <TableRow
      onClick={() => {
        selectStock(stock.symbol);
      }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      className={classNames(classes.row, hover && classes.rowHover)}
    >
      <TableCell align="center" component="th" scope="row">
        <ListItemText
          primary={
            <Typography className={classes.defaultText} color="primary">
              {stock.symbol}
            </Typography>
          }
          secondary={stock.name}
          secondaryTypographyProps={{
            color: 'textSecondary',
          }}
        />
      </TableCell>
      <TableCell align="center" className={classes.defaultText}>
        <EuroCurrency
          value={stock.price}
          decimalSeperator="."
          thousandSeperator=","
        />
      </TableCell>
      <TableCell align="center">
        <Performance
          value={stock.per7d}
          decimalSeperator="."
          thousandSeperator=","
        />
      </TableCell>
      <TableCell align="center">
        <Performance
          value={stock.per30d}
          decimalSeperator="."
          thousandSeperator=","
        />
      </TableCell>
      <TableCell align="center">
        <Typography color="primary" className={classes.defaultText}>
          {moneyFormat(stock.marketCapitalization)}
        </Typography>
      </TableCell>
      <TableCell align="center">
        <EuroCurrency
          value={stock.analystTargetPrice}
          decimalSeperator="."
          thousandSeperator=","
        />
      </TableCell>
      <TableCell align="center">
        <Valuation value={stock.valuation} />
      </TableCell>
      <TableCell align="center">
        <Performance
          value={stock.div}
          decimalSeperator="."
          thousandSeperator=","
        />
      </TableCell>
      <TableCell align="center">
        <Typography color="primary" className={classes.defaultText}>
          {t(`${stock.industry}`)}
        </Typography>
      </TableCell>
    </TableRow>
  );
};

export type DashboardTableProps = {
  stocks: API.Stock[];
  selectStock: (symbol: string) => void;
};

const DashboardTable: React.FC<DashboardTableProps> = ({
  stocks,
  selectStock,
}) => {
  const { t } = useTranslation();

  return (
    <TableContainer component={Paper}>
      <Table aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell align="center">{t('stock.name')}</TableCell>
            <TableCell align="center">{t('stock.lastPrice')}</TableCell>
            <TableCell align="center">{t('stock.7d')}</TableCell>
            <TableCell align="center">{t('stock.30d')}</TableCell>
            <TableCell align="center">{t('stock.marketCap')}</TableCell>
            <TableCell align="center">{t('stock.analystsTarget')}</TableCell>
            <TableCell align="center">{t('stock.valuation')}</TableCell>
            <TableCell align="center">{t('stock.div')}</TableCell>
            <TableCell align="center">{t('stock.industry')}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {stocks.map((s) => (
            <DashboardTableRow
              stock={s}
              selectStock={selectStock}
              key={s.symbol}
            />
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default DashboardTable;
