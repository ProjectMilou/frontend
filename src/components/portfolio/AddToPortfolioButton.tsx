import React from 'react';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Checkbox from '@material-ui/core/Checkbox';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { useTranslation } from 'react-i18next';
import * as API from '../../portfolio/APIClient';
import { errorMessageKey, errorTitleKey } from '../../Errors';
import ProgressButton from './ProgressButton';

// stylesheet for the entire add to portfolio component
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    addButton: {
      padding: '0.25rem 1rem',
      backgroundColor: '#3fbcf2',
      '&:hover': {
        backgroundColor: '#84d4f7',
      },
      whiteSpace: 'nowrap',
    },
    Dialog: {
      minWidth: '40rem',
    },
    dialogHeader: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
    // dialog header
    headerCloseButton: {
      position: 'relative',
      color: theme.palette.grey[500],
    },
    span: {
      display: 'flex',
    },
    text: {
      margin: 'auto',
    },
    // dialog body
    DialogContent: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      margin: 0,
      padding: theme.spacing(1),
    },
    // List component
    List: {
      width: '100%',
      backgroundColor: 'white',
    },
    listItem: {
      width: '100%',
      margin: '0.75rem 0',
      height: '3.25rem',
    },
    // error helper text
    helperText: {
      position: 'absolute',
      top: '100%',
      right: '-50%',
      whiteSpace: 'nowrap',
    },
    dialogFooter: {
      margin: 0,
      padding: theme.spacing(2),
      display: 'flex',
      justifyContent: 'end',
    },
  })
);

type ListEntryProps = {
  id: string;
  name: string;
  checked: CheckedType;
  handleToggle: (id: string) => () => void;
  changeAmount: (id: string, amount: string) => void;
  pattern: RegExp;
};

const ListEntry: React.FC<ListEntryProps> = ({
  id,
  name,
  checked,
  handleToggle,
  changeAmount,
  pattern,
}) => {
  const classes = useStyles();

  return (
    <ListItem role={undefined} dense className={classes.listItem}>
      <ListItemIcon>
        <Checkbox
          edge="start"
          checked={typeof checked[id] !== 'undefined'}
          tabIndex={-1}
          disableRipple
          inputProps={{
            'aria-labelledby': `checkbox-list-label-${id}`,
          }}
          color="secondary"
          onClick={handleToggle(id)}
        />
      </ListItemIcon>
      <ListItemText primary={name} />
      <TextField
        variant="outlined"
        value={checked[id]}
        error={!pattern.test(checked[id])}
        helperText={
          pattern.test(checked[id]) ? '' : 'Please enter a positive number!'
        }
        FormHelperTextProps={{
          className: classes.helperText,
        }}
        onChange={(e) => {
          changeAmount(id, e.target.value);
        }}
        size="small"
        margin="dense"
        inputProps={{
          maxLength: 5,
          style: {
            textAlign: 'center',
            paddingRight: 0,
            paddingLeft: '0.2rem',
          },
        }}
        style={{
          display: typeof checked[id] !== 'undefined' ? 'block' : 'none',
          width: '4rem',
        }}
      />
    </ListItem>
  );
};

type CheckedType = {
  // id of portfolio: qty of stock;
  [id: string]: string;
};

type AddToPortfolioButtonType = {
  isin: string;
  token: string;
};

// returns the add to portfolio button and all subcomponents including the dialog window
const AddToPortfolioButton: React.FC<AddToPortfolioButtonType> = ({
  isin,
  token,
}) => {
  // style and translation hooks
  const classes = useStyles();
  const { t } = useTranslation();
  // pattern for testing input fields
  const pattern = /^(?:[1-9]\d*|0)?(?:\.\d+)?$/;

  // the list of portfolios belonging to the user
  const [portfolios, setPortfolios] = React.useState<API.PortfolioOverview[]>(
    []
  );

  // the state that controls whether the dialog window is open or closed
  const [open, setOpen] = React.useState(false);
  // true when waiting for an API response
  const [loading, setLoading] = React.useState<boolean>(false);
  const [error, setError] = React.useState<Error | undefined>();
  // keep track of which portfolios have been selected and the amount in the textfield
  const [checked, setChecked] = React.useState<CheckedType>({});

  // defines function to fetch from api
  const isMounted = React.useRef(true);
  const fetch = async () => {
    setError(undefined);
    try {
      const p = await API.list(token);
      if (isMounted.current) {
        setPortfolios(p);
      }
    } catch (e) {
      if (isMounted.current) {
        setError(e);
      }
    }
  };

  // opon mount fetch list of portfolios from api
  React.useEffect(() => {
    fetch();
    return () => {
      isMounted.current = false;
    };
    // deps must be empty because the function should only be called on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const openDialog = () => {
    // reset the dialog's state on (re-)open
    setLoading(false);
    setError(undefined);
    setChecked({});
    setOpen(true);
  };

  // handle saving selections
  const handleSave = async () => {
    setLoading(true);

    try {
      // parse the checked state into the form expected by the api
      const modifications: API.PortfoliosQty[] = Object.entries(checked).map(
        ([id, qty]) => ({
          id,
          // TODO: use react-number-format for validation,
          //  store numbers instead of strings, avoid converting
          qty: Number.parseFloat(qty),
        })
      );
      // make api put call
      await API.stock(token, isin, modifications);
      setOpen(false);
    } catch (e) {
      // on error stop the loading animation and display error to user
      setLoading(false);
      setError(e);
    }
  };

  // handles checking and unchecking portfolios
  const handleCheckToggle = (id: string) => () => {
    const newChecked = { ...checked };

    if (typeof newChecked[id] === 'undefined') {
      // if unchecked, check it with an initial value of 1
      newChecked[id] = '1';
    } else {
      // if checked remove check
      delete newChecked[id];
    }

    setChecked(newChecked);
  };

  //  a function that sets the amount everytime a textfield is changed
  const changeAmount = (id: string, amount: string) => {
    // when a user changes the input field the checked state gets updated
    setChecked({ ...checked, [id]: amount });
  };

  return (
    <div>
      <Button
        variant="contained"
        className={classes.addButton}
        onClick={() => openDialog()}
      >
        {t('analyzer.addToPortfolio')}
      </Button>
      <Dialog
        onClose={() => setOpen(false)}
        aria-labelledby="dialog-title"
        open={open}
        className={classes.Dialog}
      >
        {/* The header of the dialog window */}
        <MuiDialogTitle disableTypography className={classes.dialogHeader}>
          <span className={classes.span}>
            <Typography variant="h6" className={classes.text}>
              {t('analyzer.dialogHeader')}
            </Typography>
          </span>
        </MuiDialogTitle>
        {/* The body of the dialog window */}
        <MuiDialogContent dividers className={classes.DialogContent}>
          <List className={classes.List} component="ol">
            {portfolios.map((portfolio) => (
              <ListEntry
                key={portfolio.id}
                id={portfolio.id}
                name={portfolio.name}
                checked={checked}
                handleToggle={handleCheckToggle}
                changeAmount={changeAmount}
                pattern={pattern}
              />
            ))}
          </List>
          {error && (
            <Typography color="error">
              <b>{t(errorTitleKey(error))}</b>: {t(errorMessageKey(error))}
            </Typography>
          )}
        </MuiDialogContent>
        {/* The footer of the dialog window */}
        <MuiDialogActions className={classes.dialogFooter}>
          <Button autoFocus onClick={() => setOpen(false)} color="primary">
            {t('portfolio.details.discardChanges')}
          </Button>
          <ProgressButton
            onClick={handleSave}
            color="primary"
            loading={loading}
            disabled={
              !Object.values(checked)
                .map((stringAmount) => pattern.test(stringAmount))
                .every((v) => v)
            }
          >
            {t('portfolio.details.saveChanges')}
          </ProgressButton>
        </MuiDialogActions>
      </Dialog>
    </div>
  );
};

// exports
export default AddToPortfolioButton;
