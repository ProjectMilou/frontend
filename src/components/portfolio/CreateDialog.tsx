import React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import NameDialog from './NameDialog';

function validateName(newName?: string): boolean {
  const n = newName?.trim();
  return !!(n && n.length);
}

export type CreateDialogProps = {
  open: boolean;
  handleClose: () => void;
  create: (name: string) => Promise<void>;
};

const CreateDialog: React.FC<CreateDialogProps> = ({
  open,
  handleClose,
  create,
}) => {
  const { t } = useTranslation();

  return (
    <NameDialog
      open={open}
      dialogTitle={t('portfolio.dialog.create.title')}
      dialogText={<Trans i18nKey="portfolio.dialog.create.text" />}
      labelKey="portfolio.name"
      actionKey="portfolio.create"
      handleClose={handleClose}
      validate={(n) => validateName(n)}
      action={async (newName) =>
        newName ? create(newName) : Promise.resolve()
      }
    />
  );
};

export default CreateDialog;
