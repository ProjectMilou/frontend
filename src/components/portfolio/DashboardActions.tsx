import React from 'react';
import { useTranslation } from 'react-i18next';
import { IconButton, makeStyles, Tooltip } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/DeleteForever';
import DuplicateIcon from '@material-ui/icons/AddToPhotos';
import EditIcon from '@material-ui/icons/Edit';
import classNames from 'classnames';
import { PortfolioOverview } from '../../portfolio/APIClient';

const useStyles = makeStyles(() => ({
  action: { display: 'inline-block' },
  disabled: {
    cursor: 'not-allowed',
  },
}));

type DashboardActionsProps = {
  portfolio: PortfolioOverview;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
};

const DashboardActions: React.FC<DashboardActionsProps> = ({
  portfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
}) => {
  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <>
      <Tooltip title={t('portfolio.rename').toString()}>
        <div className={classes.action}>
          <IconButton
            onClick={(e) => {
              renamePortfolio(portfolio.id);
              e.stopPropagation();
            }}
          >
            <EditIcon />
          </IconButton>
        </div>
      </Tooltip>
      <Tooltip title={t('portfolio.duplicate').toString()}>
        <div className={classes.action}>
          <IconButton
            onClick={(e) => {
              duplicatePortfolio(portfolio.id);
              e.stopPropagation();
            }}
          >
            <DuplicateIcon />
          </IconButton>
        </div>
      </Tooltip>
      <Tooltip
        title={(portfolio.virtual
          ? t('portfolio.delete')
          : t('portfolio.deleteReal')
        ).toString()}
        // stop click event propagation here because the delete button can be disabled
        onClick={(e) => e.stopPropagation()}
      >
        <div
          className={classNames(
            classes.action,
            !portfolio.virtual && classes.disabled
          )}
        >
          <IconButton
            disabled={!portfolio.virtual}
            onClick={() => {
              deletePortfolio(portfolio.id);
            }}
          >
            <DeleteIcon />
          </IconButton>
        </div>
      </Tooltip>
    </>
  );
};

export default DashboardActions;
