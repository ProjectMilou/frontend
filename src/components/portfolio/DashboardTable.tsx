import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  Button,
  lighten,
  ListItemText,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  TableSortLabel,
  Theme,
  Typography,
} from '@material-ui/core';
import classNames from 'classnames';
import {
  NonEmptyPortfolioOverview,
  PortfolioOverview,
} from '../../portfolio/APIClient';
import DashboardActions from './DashboardActions';
import EuroCurrency from '../shared/EuroCurrency';
import Performance from '../shared/Performance';

const useStyles = makeStyles((theme: Theme) => ({
  action: { display: 'inline-block' },
  row: {
    cursor: 'pointer',
  },
  rowHover: {
    backgroundColor: lighten(theme.palette.primary.light, 0.85),
  },
  positionCount: {
    fontSize: '24px',
  },
  disabled: {
    cursor: 'not-allowed',
  },
  createButton: {
    marginTop: '25px',
  },
}));

type DashboardTableRowProps = {
  portfolio: PortfolioOverview;
  selectPortfolio: (id: string) => void;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
};

const DashboardTableRow: React.FC<DashboardTableRowProps> = ({
  portfolio,
  selectPortfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
}) => {
  const [hover, setHover] = React.useState<boolean>(false);

  const { t } = useTranslation();
  const classes = useStyles();

  return (
    <TableRow
      onClick={() => {
        selectPortfolio(portfolio.id);
      }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      className={classNames(classes.row, hover && classes.rowHover)}
    >
      <TableCell align="center">
        {'score' in portfolio && portfolio.score}
      </TableCell>
      <TableCell align="center" component="th" scope="row">
        <ListItemText
          primary={<Typography color="primary">{portfolio.name}</Typography>}
          secondary={
            portfolio.virtual ? t('portfolio.virtual') : t('portfolio.real')
          }
          secondaryTypographyProps={{
            color: portfolio.virtual ? 'textSecondary' : 'secondary',
          }}
        />
      </TableCell>
      <TableCell align="center" className={classes.positionCount}>
        {portfolio.positionCount}
      </TableCell>
      <TableCell align="center">
        <EuroCurrency value={portfolio.value} />
      </TableCell>
      <TableCell align="center">
        <Performance value={portfolio.perf7d} />
      </TableCell>
      <TableCell align="center">
        <Performance value={portfolio.perf1y} />
      </TableCell>
      <TableCell align="center">
        <DashboardActions
          portfolio={portfolio}
          renamePortfolio={renamePortfolio}
          duplicatePortfolio={duplicatePortfolio}
          deletePortfolio={deletePortfolio}
        />
      </TableCell>
    </TableRow>
  );
};

type Order = 'asc' | 'desc';

type KeysOfUnion<T> = T extends T ? keyof T : never;
type PortfolioOverviewKeys = KeysOfUnion<PortfolioOverview>;

interface SortableHeadCell {
  id: PortfolioOverviewKeys;
  label: string;
  numeric: boolean;
}

export type DashboardTableHeadProps = {
  onRequestSort: (property: PortfolioOverviewKeys) => void;
  order: Order;
  orderBy: PortfolioOverviewKeys;
};

export const DashboardTableHead: React.FC<DashboardTableHeadProps> = ({
  onRequestSort,
  order,
  orderBy,
}) => {
  const { t } = useTranslation();
  const headCells: SortableHeadCell[] = [
    { id: 'score', numeric: true, label: t('portfolio.score') },
    { id: 'name', numeric: false, label: t('portfolio.name') },
    {
      id: 'positionCount',
      numeric: true,
      label: t('portfolio.positionsCount'),
    },
    { id: 'value', numeric: true, label: t('portfolio.value') },
    { id: 'perf7d', numeric: true, label: t('portfolio.7d') },
    { id: 'perf1y', numeric: true, label: t('portfolio.1y') },
  ];

  return (
    <TableHead>
      <TableRow>
        {headCells.map((headCell) => (
          <TableCell
            align="center"
            key={headCell.id}
            sortDirection={orderBy === headCell.id ? order : false}
          >
            <TableSortLabel
              active={orderBy === headCell.id}
              direction={orderBy === headCell.id ? order : 'asc'}
              onClick={() => onRequestSort(headCell.id)}
            >
              {headCell.label}
            </TableSortLabel>
          </TableCell>
        ))}
        <TableCell align="center">{t('portfolio.actions')}</TableCell>
      </TableRow>
    </TableHead>
  );
};

export type DashboardTableProps = {
  portfolios: PortfolioOverview[];
  selectPortfolio: (id: string) => void;
  renamePortfolio: (id: string) => void;
  duplicatePortfolio: (id: string) => void;
  deletePortfolio: (id: string) => void;
  createPortfolio: () => void;
};

function comparatorTable<T>(a: T, b: T) {
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
}

function isDefined(obj: Record<string, unknown>, key: string) {
  return key in obj && obj[key] !== undefined;
}

function sortPortfolios(
  portfolios: PortfolioOverview[],
  orderBy: PortfolioOverviewKeys,
  order: Order
) {
  // portfolios without undefined values in the sorting column
  const nonUndefined = portfolios.filter((values) =>
    isDefined(values, orderBy)
  ) as NonEmptyPortfolioOverview[];

  const sorted =
    order === 'asc'
      ? nonUndefined.sort((a, b) => comparatorTable(a[orderBy], b[orderBy]))
      : nonUndefined.sort((a, b) => comparatorTable(b[orderBy], a[orderBy]));

  // undefined values are always at the end
  return [
    ...sorted,
    ...portfolios.filter((values) => !isDefined(values, orderBy)),
  ];
}

const DashboardTable: React.FC<DashboardTableProps> = ({
  portfolios,
  selectPortfolio,
  renamePortfolio,
  duplicatePortfolio,
  deletePortfolio,
  createPortfolio,
}) => {
  const [order, setOrder] = React.useState<Order>('desc');
  const [orderBy, setOrderBy] = React.useState<PortfolioOverviewKeys>('score');
  const [sortedPortfolios, setSortedPortfolios] = React.useState(() =>
    sortPortfolios(portfolios, orderBy, order)
  );

  // only sort portfolios when needed
  React.useEffect(() => {
    setSortedPortfolios(sortPortfolios(portfolios, orderBy, order));
  }, [portfolios, orderBy, order]);

  const classes = useStyles();
  const { t } = useTranslation();

  const handleRequestSort = (property: PortfolioOverviewKeys) => {
    const isAsc = orderBy === property && order === 'asc';
    setOrder(isAsc ? 'desc' : 'asc');
    setOrderBy(property);
  };
  // TODO: Improve portfolio score visualization
  return (
    <>
      {!!portfolios.length && (
        <TableContainer component={Paper}>
          <Table>
            <DashboardTableHead
              onRequestSort={handleRequestSort}
              order={order}
              orderBy={orderBy}
            />
            <TableBody>
              {sortedPortfolios.map((p) => (
                <DashboardTableRow
                  portfolio={p}
                  selectPortfolio={selectPortfolio}
                  key={p.id}
                  renamePortfolio={renamePortfolio}
                  duplicatePortfolio={duplicatePortfolio}
                  deletePortfolio={deletePortfolio}
                />
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )}
      <Button
        className={classes.createButton}
        variant="outlined"
        color="primary"
        onClick={() => createPortfolio()}
      >
        {t('portfolio.dashboard.createPortfolio')}
      </Button>
    </>
  );
};

export default DashboardTable;
