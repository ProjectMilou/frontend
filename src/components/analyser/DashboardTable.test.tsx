import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import * as API from '../../portfolio/APIClient';
import DashboardTable, { DashboardTableProps } from './DashboardTable';

describe('DashboardTable', () => {
  const defaultProps: DashboardTableProps = {
    portfolios: [API.MockOverview, API.MockOverviewTwo],
    selectPortfolio: jest.fn(),
  };

  const renderComponent = (newProps?: Partial<DashboardTableProps>) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(<DashboardTable {...props} />),
      props,
    };
  };

  test('display portfolio information', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('select portfolio on click', () => {
    const { props, getByText } = renderComponent();
    fireEvent.click(getByText(props.portfolios[0].name));
    expect(props.selectPortfolio).toHaveBeenCalledWith(props.portfolios[0].id);
    fireEvent.click(getByText(props.portfolios[1].name));
    expect(props.selectPortfolio).toHaveBeenCalledWith(props.portfolios[1].id);
  });
});
