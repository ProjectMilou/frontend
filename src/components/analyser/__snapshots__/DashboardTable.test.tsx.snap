// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardTable display portfolio information 1`] = `
<div>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      aria-label="simple table"
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.score
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.positionsCount
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.value
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.7d
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            portfolio.1y
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root makeStyles-row-2"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          />
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            role="cell"
            scope="row"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
              >
                test
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                portfolio.virtual
              </p>
            </div>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-positionCount-4 MuiTableCell-alignCenter"
          >
            3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-euroCurrency-6 makeStyles-euroCurrency-7"
            >
              9,00 €
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-performance-8 makeStyles-performance-9"
            >
              -1,23 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-performance-8 makeStyles-performance-10"
            >
              13,37 %
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root makeStyles-row-2"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            20
          </td>
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
            role="cell"
            scope="row"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary"
              >
                testTwo
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorSecondary MuiTypography-displayBlock"
              >
                portfolio.real
              </p>
            </div>
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body makeStyles-positionCount-4 MuiTableCell-alignCenter"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-euroCurrency-6 makeStyles-euroCurrency-11"
            >
              3,00 €
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-performance-8 makeStyles-performance-12"
            >
              0,00 %
            </span>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <span
              class="makeStyles-performance-8 makeStyles-performance-13"
            >
              -1,00 %
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
